### PR TITLE
Add support for bidirectional mapping (output mapper)

### PR DESCRIPTION
**Pull request by JanTvrdik**
## Bidirectional Mapping Support (Output Mapper)

This PR extends the library to support the inverse mapping direction: `T → mixed` (output/serialization), in addition to the existing `mixed → T` (input/deserialization).

### Breaking Changes

- **`MapperProvider::get()` removed** — replaced by `getInputMapper()` and `getOutputMapper()`
- **`Mapper<T>` now `Mapper<I, O>`** — contravariant input, covariant output (two generic parameters)
- **Attributes moved** to `Compiler\Attribute\` namespace (e.g. `MapInt`, `MapString`, `MapObject`, etc.)
- **Attributes no longer implement `MapperCompiler`** — they now implement `InputMapperCompilerProvider` and/or `OutputMapperCompilerProvider` (or the combined `MapperCompilerProvider`)
- **`*MapperCompiler` classes renamed** with `Input` prefix (e.g. `ArrayShapeMapperCompiler` → `ArrayShapeInputMapperCompiler`) and moved to `Compiler\Mapper\Input\` namespace
- **`DelegateMapperCompiler` split** into `DelegateInputMapperCompiler` / `DelegateOutputMapperCompiler` extending `AbstractDelegateMapperCompiler`
- **`DefaultMapperCompilerFactory::create()` renamed** to `createProvider()` returning `MapperCompilerProvider` (which provides both input and output compilers)

### New Features

- **`OutputMapper<T>`** — new runtime interface (`Mapper<T, mixed>`) for serializing objects to plain data
- **Output mapper compilers** — `ObjectOutputMapperCompiler`, `ArrayShapeOutputMapperCompiler`, `ListOutputMapperCompiler`, `DiscriminatedObjectOutputMapperCompiler`, `EnumOutputMapperCompiler`, `DateTimeImmutableOutputMapperCompiler`, `NullableOutputMapperCompiler`, `OptionalOutputMapperCompiler`, `DelegateOutputMapperCompiler`
- **`PassthroughMapperCompiler`** — no-op compiler for primitives that need no transformation, used to skip unnecessary work in output mapping

### Major Internal Changes

- `PhpCodeBuilder::importType()` now returns a cloned `TypeNode` instead of mutating input
- Output mapper code generation with optimizations: passthrough detection, expression inlining, sealed array shape return types

### Test plan

- Round-trip tests verify `data → object → data` identity for all supported types
- Dedicated output mapper compiler tests with snapshot assertions
- Error-path tests: discriminated output with unrecognized subtype, edge values (zero, false, empty string)

### CODE REVIEW

File: src/Compiler/MapperFactory/DefaultMapperCompilerFactory.php [hunk 1]
```diff
@@ -434,33 +497,41 @@ protected function getConstructorParameterTypes(
...
+        foreach ($parameterReflection->getAttributes(InputMapperCompilerProvider::class, ReflectionAttribute::IS_INSTANCEOF) as $attribute) {
+            $inputProviders[] = $attribute->newInstance();
         }
 
         foreach ($parameterReflection->getAttributes(ValidatorCompiler::class, ReflectionAttribute::IS_INSTANCEOF) as $attribute) {
             $validators[] = $attribute->newInstance();
         }
 
-        $mapper = match (count($mappers)) {
+        /** @var MapperCompilerProvider $provider */
```
Comment (janedbal) [inline src/Compiler/MapperFactory/DefaultMapperCompilerFactory.php]: This is not guarranteed for `$inputProviders[0]` arm, not sure it can be somehow.

File: src/Compiler/Mapper/Output/DiscriminatedObjectOutputMapperCompiler.php [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+use ShipMonk\InputMapper\Compiler\Php\PhpCodeBuilder;
+use ShipMonk\InputMapper\Compiler\Type\GenericTypeParameter;
+use ShipMonk\InputMapper\Runtime\Exception\MappingFailedException;
+use function count;
+use function ucfirst;
+
+/**
+ * @template T of object
+ */
+class DiscriminatedObjectOutputMapperCompiler implements GenericMapperCompiler
```
Comment (janedbal) [inline src/Compiler/Mapper/Output/DiscriminatedObjectOutputMapperCompiler.php]: AI hint: The input counterpart validates PhpDocTypeUtils::isSubTypeOf(). The output version has zero such validation — no imports of PhpDocTypeUtils or CannotCompileMapperException. A misconfigured subtype silently generates code that never matches.

File: src/Compiler/Attribute/MapValidated.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    public function getInputMapperCompiler(): MapperCompiler
+    {
+        return new ValidatedInputMapperCompiler($this->mapperCompilerProvider->getInputMapperCompiler(), $this->validatorCompilers);
+    }
+
+    public function getOutputMapperCompiler(): MapperCompiler
+    {
+        return $this->mapperCompilerProvider->getOutputMapperCompiler();
```
[the remark this answers is not in the document]
Comment (JanTvrdik) [reply src/Compiler/Attribute/MapValidated.php:35]: yes
